diff -uNr char.orig/Makefile char/Makefile
--- char.orig/Makefile	Fri Oct 15 15:12:07 1999
+++ char/Makefile	Fri Oct 15 15:45:54 1999
@@ -343,12 +343,14 @@
 endif
 
 ifeq ($(CONFIG_VIDEO_BT848),y)
-L_OBJS += bttv.o msp3400.o tuner.o
+L_OBJS += msp3400.o tuner.o
 L_I2C=y
+LX_OBJS += bttv.o
 else
   ifeq ($(CONFIG_VIDEO_BT848),m)
-    M_OBJS += bttv.o msp3400.o tuner.o
+    M_OBJS += msp3400.o tuner.o
     M_I2C=y
+    MX_OBJS += bttv.o
   endif
 endif
 
diff -uNr char.orig/bttv.c char/bttv.c
--- char.orig/bttv.c	Fri Oct 15 15:12:07 1999
+++ char/bttv.c	Fri Oct 15 15:45:54 1999
@@ -264,6 +264,41 @@
 	return 0;
 }
 
+/* ----------------------------------------------------------------------- */
+/* GPIO functions (for use by lirc drivers)                                */
+
+int read_bt848_gpio(int card, int latch, unsigned long mask, unsigned long *val)
+{
+	struct bttv *btv;
+	
+	if(card >= BTTV_MAX)
+		return -1;
+
+	btv = &bttvs[card];
+
+	btand(~(latch? 0ul: mask), BT848_GPIO_REG_INP);
+
+	*val = (btread(BT848_GPIO_DATA) & mask);
+
+	return 0;
+}
+
+int write_bt848_gpio(int card, int latch, unsigned long mask, unsigned long val)
+{
+	struct bttv *btv;
+	
+	if(card >= BTTV_MAX)
+		return -1;
+
+	btv = &bttvs[card];
+
+	btand(~(latch? 0ul: mask), BT848_GPIO_REG_INP);
+
+	btaor(val & mask, ~mask, BT848_GPIO_DATA);
+
+	return 0;
+}
+
 
 /* ----------------------------------------------------------------------- */
 /* I2C functions                                                           */
@@ -550,7 +585,7 @@
 	/* Diamond DTV2000 */
         { 3, 1, 0, 2, 3, { 2, 3, 1, 1}, { 0, 1, 0, 1, 3}},
 	/* AVerMedia TVPhone */
-        { 3, 1, 0, 3,15, { 2, 3, 1, 1}, {12, 0,11,11, 0}},
+        { 3, 1, 0, 3,15, { 2, 3, 1, 1}, {12, 4,11,11, 0}},
         /* Matrix Vision MV-Delta */
         { 5, 1, -1, 3, 0, { 2, 3, 1, 0, 0}},
         /* Fly Video II */
@@ -3893,7 +3928,8 @@
 
 #ifdef MODULE
 
-EXPORT_NO_SYMBOLS;
+EXPORT_SYMBOL(read_bt848_gpio);
+EXPORT_SYMBOL(write_bt848_gpio);
 
 int init_module(void)
 {
