diff -u --recursive linux/drivers/char/Config.in linux-irctl/drivers/char/Config.in
--- linux/drivers/char/Config.in	Fri May  7 20:05:30 1999
+++ linux-irctl/drivers/char/Config.in	Mon Jul 26 22:14:40 1999
@@ -132,6 +132,9 @@
   if [ "$CONFIG_PCI" != "n" ]; then
     dep_tristate 'BT848 Video For Linux' CONFIG_VIDEO_BT848 $CONFIG_VIDEO_DEV
   fi
+  if [ "$CONFIG_VIDEO_BT848" != "n" ]; then
+    bool 'FlyVideo98 Remote Control' CONFIG_BTTV_REMOTE
+  fi
   if [ "$CONFIG_PARPORT" != "n" ]; then
     dep_tristate 'Quickcam BW Video For Linux' CONFIG_VIDEO_BWQCAM $CONFIG_VIDEO_DEV $CONFIG_PARPORT
     dep_tristate 'Colour QuickCam Video For Linux (EXPERIMENTAL)' CONFIG_VIDEO_CQCAM $CONFIG_VIDEO_DEV $CONFIG_PARPORT
diff -u --recursive linux/drivers/char/Makefile linux-irctl/drivers/char/Makefile
--- linux/drivers/char/Makefile	Fri May  7 20:05:30 1999
+++ linux-irctl/drivers/char/Makefile	Sun Aug  1 11:13:26 1999
@@ -317,6 +317,9 @@
   ifeq ($(CONFIG_VIDEO_BT848),m)
     M_OBJS += bttv.o msp3400.o tuner.o
     M_I2C=y
+    ifeq ($(CONFIG_BTTV_REMOTE),m)
+	CFLAGS_bttv.o += -DBTTV_GPIOMON
+    endif
   endif
 endif
 
diff -u --recursive linux/drivers/char/bttv.c linux-irctl/drivers/char/bttv.c
--- linux/drivers/char/bttv.c	Tue Jun  8 01:18:17 1999
+++ linux-irctl/drivers/char/bttv.c	Sun Aug  1 11:16:43 1999
@@ -20,6 +20,11 @@
     Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
 
+#include <linux/config.h>
+#ifndef EXPORT_SYMTAB
+#define EXPORT_SYMTAB
+#endif
+
 #include <linux/module.h>
 #include <linux/version.h>
 #include <linux/delay.h>
@@ -119,6 +124,10 @@
 static int bttv_num;			/* number of Bt848s in use */
 static struct bttv bttvs[BTTV_MAX];
 
+#ifdef BTTV_GPIOMON
+static bttv_gpio_monitor_t gpio_monitor=NULL;
+#endif
+
 #define I2C_TIMING (0x7<<4)
 #define I2C_DELAY   10
 
@@ -548,6 +557,8 @@
         { 3, 1, 0, 2, 0x8300f8, { 2, 3, 1, 1,0}, {0x4fa007,0xcfa007,0xcfa007,0xcfa007,0xcfa007,0xcfa007}},
         /* AVEC Intercapture */
         { 3, 1, 9, 2, 0, { 2, 3, 1, 1}, { 0, 0, 0, 0, 0}},
+	/* Fly Video 98 */
+	{ 3, 1, 0, 2, 0x8dfe00, {2, 3, 1, 1}, {0, 0x8dfe00, 0x800, 0x400, 0x8dfe00, 0 }},
 };
 #define TVCARDS (sizeof(tvcards)/sizeof(tvcard))
 
@@ -3157,6 +3168,18 @@
 		bt848_dma(btv, 0);
 }
 
+
+
+#ifdef BTTV_GPIOMON
+bttv_gpio_monitor_t bttv_set_gpio_monitor(bttv_gpio_monitor_t callback)
+{
+        bttv_gpio_monitor_t prev=gpio_monitor;
+  
+        gpio_monitor=callback;
+        return prev;
+}
+#endif
+
 static int init_bt848(int i)
 {
         struct bttv *btv = &bttvs[i];
@@ -3270,15 +3293,34 @@
 	btwrite(0xfffffUL, BT848_INT_STAT);
         
 	/* set interrupt mask */
+#ifdef BTTV_GPIOMON
+	/* set rising edge of inverted GPINTR pin as irq trigger */
+	btwrite(btread(BT848_GPIO_DMA_CTL)
+    	    |BT848_GPIO_DMA_CTL_GPINTC
+    	    ,BT848_GPIO_DMA_CTL);
+	btwrite(btread(BT848_GPIO_DMA_CTL)
+	    |BT848_GPIO_DMA_CTL_GPINTI
+	    ,BT848_GPIO_DMA_CTL);
+	/* and irq mask */    
 	btwrite(btv->triton1|
                 /*BT848_INT_PABORT|BT848_INT_RIPERR|BT848_INT_PPERR|
                   BT848_INT_FDSR|BT848_INT_FTRGT|BT848_INT_FBUS|*/
+                BT848_INT_GPINT|
                 BT848_INT_VSYNC|
 		BT848_INT_SCERR|
 		BT848_INT_RISCI|BT848_INT_OCERR|BT848_INT_VPRES|
 		BT848_INT_FMTCHG|BT848_INT_HLOCK,
 		BT848_INT_MASK);
-
+#else
+	btwrite(btv->triton1|
+                /*BT848_INT_PABORT|BT848_INT_RIPERR|BT848_INT_PPERR|
+                  BT848_INT_FDSR|BT848_INT_FTRGT|BT848_INT_FBUS|*/
+                BT848_INT_VSYNC|
+		BT848_INT_SCERR|
+		BT848_INT_RISCI|BT848_INT_OCERR|BT848_INT_VPRES|
+		BT848_INT_FMTCHG|BT848_INT_HLOCK,
+		BT848_INT_MASK);
+#endif
 	make_vbitab(btv);
 	bt848_set_risc_jmps(btv);
   
@@ -3336,6 +3378,17 @@
 		/* get device status bits */
 		dstat=btread(BT848_DSTATUS);
     
+#ifdef BTTV_GPIOMON
+                if ( astat & BT848_INT_GPINT )
+        	{
+            		unsigned long b; 
+                    
+                	b=btread(BT848_GPIO_DATA);
+                    	IDEBUG(printk("bttv%d: GPIO: %04lx INT_STAT: %04lx \n", 
+                    		btv->nr, b&0x1ff, (long)stat));
+            		(gpio_monitor) ? gpio_monitor(btv->nr, b) : 0;
+		}
+#endif
 		if (astat&BT848_INT_FMTCHG) 
 		{
 			IDEBUG(printk ("bttv%d: IRQ_FMTCHG\n", btv->nr));
@@ -3840,9 +3893,17 @@
 	}
 }
 
-#ifdef MODULE
-
+#ifdef BTTV_GPIOMON
+#if LINUX_VERSION_CODE >= 0x020100
+EXPORT_SYMBOL(bttv_set_gpio_monitor);
+#endif
+#else
 EXPORT_NO_SYMBOLS;
+#endif
+
+
+
+#ifdef MODULE
 
 int init_module(void)
 {
diff -u --recursive linux/drivers/char/bttv.h linux-irctl/drivers/char/bttv.h
--- linux/drivers/char/bttv.h	Tue Jun  8 01:18:17 1999
+++ linux-irctl/drivers/char/bttv.h	Sun Aug  1 11:12:10 1999
@@ -164,6 +165,11 @@
 	int i2c_command;
 	int triton1;
 };
+#endif
+
+#ifdef BTTV_GPIOMON
+typedef void (*bttv_gpio_monitor_t)(int nr, unsigned long data);
+extern bttv_gpio_monitor_t bttv_set_gpio_monitor(bttv_gpio_monitor_t callback);
 #endif
 
 /*The following should be done in more portable way. It depends on define
